### PR TITLE
Adding synchronization and other features to extended test cluster loader.

### PR DESCRIPTION
**Pull request by jmencak**
Fixes:
- number of templates no longer ignored
- tuningsets no longer ignored for templates
- maxRetries reflected the number of all tries, not retries

New features:
- label-based pod post deployment synchronization at pod, template
  or global level
- creation of config maps and secrets from files
- support for getting parameter values from the environment

### CODE REVIEW

**Comment by jmencak** [comment]
@sjug creating this in parallel with https://github.com/openshift/origin/pull/17072 for review purposes, while you're installing go ;-)

**Comment by jmencak** [comment]
Flakes on:
  * "ci/openshift-jenkins/cmd"
    test/cmd/observe.sh:34: executing 'oc observe services --once --all-namespaces -a "bad{ .metadata.annotations.unset }key" --strict-templates' expecting failure and text 'annotations is not found': the output content test failed
  * ci/openshift-jenkins/extended_conformance_install 
    rm: cannot remove â€˜/tmp/etcd/memberâ€™: Permission denied

**Comment by jmencak** [comment]
Flakes on: 
  *  ci/openshift-jenkins/extended_conformance_gce
  `rm: cannot remove ‘/tmp/etcd’: Operation not permitted`
  *  ci/openshift-jenkins/integration
  `error retrieving resource lock kube-system/kube-controller-manager: Get https://127.0.0.1:21143/api/v1/namespaces/kube-system/configmaps/kube-controller-manager: dial tcp 127.0.0.1:21143: getsockopt: connection refused`

**Comment by jmencak** [comment]
@sjug PTAL

File: test/extended/cluster/utils.go [hunk 1]
```diff
@@ -178,13 +252,49 @@ func convertVariablesToMap(params map[string]interface{}) map[string]string {
...
 		k = strings.ToUpper(k)
-		if v != 0 && v != "" {
-			args = append(args, "-p")
-			args = append(args, fmt.Sprintf("%s=%v", k, v))
+		val := v
+		if v == nil {
+			// Parameter not defined, see if it is defined in the environment.
+			var found bool
+			val, found = os.LookupEnv(fmt.Sprintf("%s", k))
+			if found == false {
```
Comment (sjug) [inline test/extended/cluster/utils.go]: `found` is boolean so we don't need to explicitly compare it.
```
if found {
    ...
```

File: test/extended/cluster/utils.go [hunk 2]
```diff
@@ -178,13 +252,49 @@ func convertVariablesToMap(params map[string]interface{}) map[string]string {
...
+				// Parameter not defined in the environment, do not define it
```
Comment (sjug) [inline test/extended/cluster/utils.go]: Logically reads nicely but we're not doing anything if it's not found.

File: test/extended/cluster/cl.go [hunk 3]
```diff
@@ -81,30 +81,34 @@ var _ = g.Describe("[Feature:Performance][Serial][Slow] Load cluster", func() {
...
 
-					if template.Parameters == nil {
-						e2e.Logf("Template environment variables will not be modified.")
-					} else {
-						params := convertVariablesToString(template.Parameters)
-						allArgs = append(allArgs, params...)
+				// Create secrets
+				if p.Secrets != nil {
+					// Secrets defined
+					for k, v := range p.Secrets {
```
Comment (sjug) [inline test/extended/cluster/cl.go]: This block looks really similar to the configmap version L87-91 😄 
Can we make it more generic too?

File: test/extended/cluster/utils.go [hunk 4]
```diff
@@ -178,13 +252,49 @@ func convertVariablesToMap(params map[string]interface{}) map[string]string {
...
+		args = append(args, "configmap")
+		args = append(args, k)
+		args = append(args, getFromFileArg(k, v))
+	} else {
+		return nil, fmt.Errorf("No or empty value provided for configmap filename.")
+	}
+	return args, err
+}
+
+func getSecretArgs(k string, v interface{}) (args []string, err error) {
```
Comment (sjug) [inline test/extended/cluster/utils.go]: nit, very similar to `getConfigMapArgs`.

File: test/extended/cluster/utils.go [hunk 5]
```diff
@@ -41,12 +48,58 @@ func ParsePods(jsonFile string) (configStruct kapiv1.Pod) {
 	return
 }
 
+// Wait for pods to go into Running or "not Running" state
+func SyncPods(c kclientset.Interface, ns string, selectors map[string]string, timeout time.Duration, negate bool) (err error) {
```
Comment (sjug) [inline test/extended/cluster/utils.go]: This has already been implemented a bunch of times in the e2e framework ([example](https://github.com/kubernetes/perf-tests/blob/master/clusterloader/framework/pod_utils.go#L47-L64)). Do we want another version here?

File: test/extended/cluster/utils.go [hunk 6]
```diff
@@ -120,19 +190,23 @@ func Server(c *PodCount) error {
...
+	if awaitShutdown {
+		select {
+		case <-c.Shutdown:
+			fmt.Println("Shutdown server")
+			ln.Close()
+			return err
+		}
 	}
+
+	return err
```
Comment (sjug) [inline test/extended/cluster/utils.go]: Shouldn't this be `return nil` or what error is this?

File: test/extended/cluster/cl.go [hunk 7]
```diff
@@ -122,11 +126,32 @@ var _ = g.Describe("[Feature:Performance][Serial][Slow] Load cluster", func() {
...
-					CreatePods(c, pod.Basename, nsName, labels, config.Spec, pod.Number, tuning)
+					err := CreatePods(c, pod.Basename, nsName, labels, config.Spec, pod.Number, tuning, &pod.Sync)
+					o.Expect(err).NotTo(o.HaveOccurred())
 				}
 			}
 		}
 
+		if sync.Running {
+			for _, ns := range namespaces {
+				err := SyncRunningPods(c, ns, sync.Selectors, time.Duration(sync.Timeout)*time.Second)
```
Comment (sjug) [inline test/extended/cluster/cl.go]:
```
time.Duration(sync.Timeout)*time.Second
```

This kind of pattern has happened a lot in here, I suggest using `time.ParseDuration` rather than hardcoding. This is something I've done in the [perf-tests cluster loader version](https://github.com/kubernetes/perf-tests/tree/master/clusterloader).

File: test/extended/cluster/context.go [hunk 8]
```diff
@@ -33,9 +36,22 @@ type ClusterLoaderObjectType struct {
 	Image      string
 	Basename   string
 	File       string
+	Sync       SyncObjectType
```
[the remark this answers is not in the document]
Comment (jmencak) [reply test/extended/cluster/context.go]: In order to perform synchronization both on template/pod and cluster-loader wide level.

File: test/extended/cluster/utils.go (hunk 5 above)
Comment (jmencak) [reply test/extended/cluster/utils.go]: So how do you implement waiting for a pod to go into anything but "Running" state using the code you pointed out?

File: test/extended/cluster/context.go [hunk 9]
```diff
@@ -13,17 +13,20 @@ type ContextType struct {
...
 type ClusterLoaderType struct {
-	Number    int `mapstructure:"num"`
-	Basename  string
-	Tuning    string
-	Pods      []ClusterLoaderObjectType
-	Templates []ClusterLoaderObjectType
+	Number     int `mapstructure:"num"`
+	Basename   string
+	Tuning     string
+	Configmaps map[string]interface{}
```
[the remark this answers is not in the document]
Comment (jmencak) [reply test/extended/cluster/context.go]: Not sure where, formatted by gofmt, also used go tool vet

**Comment by jmencak** [comment]
@sjug please check that all your concerns have been addressed.

File: test/extended/cluster/context.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (sjug) [reply test/extended/cluster/context.go]: The columns should align, not sure why you're not getting proper output.

File: test/extended/cluster/context.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (sjug) [reply test/extended/cluster/context.go]: We need to discuss this further.

File: test/extended/cluster/utils.go (hunk 5 above)
Comment (sjug) [reply test/extended/cluster/utils.go]: Usually we would expect a specific state.

File: test/extended/cluster/utils.go [hunk 10]
```diff
@@ -106,12 +181,15 @@ func GetTuningSet(tuningSets []TuningSetType, podTuning string) (tuning *TuningS
 	return nil
 }
 
-// Server is the webservice that will syncronize the start and stop of Pods
-func Server(c *PodCount) error {
+// Server is the webservice that will synchronize the start and stop of Pods
+func Server(c *PodCount, Port int, awaitShutdown bool) error {
```
Comment (sjug) [inline test/extended/cluster/utils.go]: `Port` should not be capitalized.

File: test/extended/cluster/utils.go [hunk 11]
```diff
@@ -106,12 +181,15 @@ func GetTuningSet(tuningSets []TuningSetType, podTuning string) (tuning *TuningS
...
 	http.HandleFunc("/start", handleStart(startHandler, c))
 	http.HandleFunc("/stop", handleStop(stopHandler, c))
+	if Port <= 0 {
```
Comment (sjug) [inline test/extended/cluster/utils.go]: How could `Port` be < 0?

File: test/extended/cluster/utils.go [hunk 12]
```diff
@@ -7,23 +7,30 @@ import (
...
 	exutil "github.com/openshift/origin/test/extended/util"
 )
 
-const maxRetries = 5
+const (
+	// The number of times we re-try to create a pod
+	maxRetries = 4
+
+	// Poll every two seconds
+	Poll = 2 * time.Second
```
Comment (sjug) [inline test/extended/cluster/utils.go]: Not sure we need to export this?

File: test/extended/cluster/utils.go (hunk 11 above)
Comment (jmencak) [reply test/extended/cluster/utils.go]: Perhaps because it is a signed int and user defined it as negative?  Most code I've seen in origin uses int for ports.

File: test/extended/cluster/context.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (jmencak) [reply test/extended/cluster/context.go]: Which columns do not align?  I do not see any mis-alignment.  Again, the code is formatted with "gofmt -s -d context.go" produces no diff. Please be specific.

File: test/extended/cluster/utils.go (hunk 5 above)
Comment (jmencak) [reply test/extended/cluster/utils.go]: I do not see why.  Expecting anything but "not Running" instead of "Complete/Error/..." is a valid reason for synchronization.

File: test/extended/cluster/utils.go (hunk 11 above)
Comment (sjug) [reply test/extended/cluster/utils.go]: I'd rather it just throw an error in that edge case, but this will be fine, maybe add a log message?

File: test/extended/cluster/context.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (sjug) [reply test/extended/cluster/context.go]: Not currently used but may be helpful in the future.

File: test/extended/cluster/context.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (sjug) [reply test/extended/cluster/context.go]: The one I've tagged, line 26?
`Configmaps` is aligned,  but `map[string]interface{}` is not.

[the remark this answers is not in the document]
Comment (sjug) [reply test/extended/cluster/context.go]: It's just on github... ffs

[the remark this answers is not in the document]
Comment (jmencak) [reply test/extended/cluster/context.go]: Great, was getting desperate...

**Comment by jmencak** [comment]
@sjug AFAIK all requests for change addressed.

**Comment by knobunc** [comment]
/lgtm
@rajatchopra PTAL

**Comment by jmencak** [comment]
Flake on ci/openshift-jenkins/extended_conformance_gce   "dial tcp 10.142.0.5:10250: getsockopt: connection refused"
